engine/access/state_stream/backend/backend_executiondata2_test.go

```diff
@@ -0,0 +1,1065 @@
+	s.snapshot.On("SealedResult").Return(s.executionResults[0], nil, nil).Maybe() // seal is not used
+	s.snapshot.On("Identities", mock.Anything).Return(s.fixedExecutionNodes, nil).Maybe()
+
+	s.params = protocolmock.NewParams(s.T())
+	s.params.On("SporkRootBlockHeight").Return(s.sporkRootBlock.Height, nil).Maybe()
+	s.params.On("SporkRootBlock").Return(s.sporkRootBlock, nil).Maybe()
+
+	s.state = protocolmock.NewState(s.T())
+	s.state.On("AtBlockID", mock.Anything).Return(s.snapshot, nil).Maybe()
+	s.state.On("Sealed").Return(s.snapshot, nil).Maybe()
+	s.state.On("Params").Return(s.params).Maybe()
+	s.state.On("Final").Return(s.snapshot, nil).Maybe()
+
+	s.headers = storagemock.NewHeaders(s.T())
+
+	// execution data stuff
+	s.executionDataBroadcaster = engine.NewBroadcaster()
+	s.executionDataTrackerReal = tracker.NewExecutionDataTracker(
+		s.log,
+		s.state,
+		s.sporkRootBlock.Height,
+		s.headers,
+		s.executionDataBroadcaster,
+		s.sporkRootBlock.Height,
+	)
+
+	s.executionDataTracker = trackermock.NewExecutionDataTracker(s.T())
+
+	s.subscriptionFactory = subscription.NewSubscriptionHandler(
+		s.log,
+		s.executionDataBroadcaster,
+		subscription.DefaultSendTimeout,
+		subscription.DefaultResponseLimit,
+		subscription.DefaultSendBufferSize,
+	)
+
+	// optimistic sync stuff
+	s.executionResultProvider = osyncmock.NewExecutionResultInfoProvider(s.T())
+	s.executionDataSnapshot = osyncmock.NewSnapshot(s.T())
+	s.executionStateCache = osyncmock.NewExecutionStateCache(s.T())
+	s.criteria = optimistic_sync.DefaultCriteria
+
+	executionNodeSelector := execution_result.NewExecutionNodeSelector(
+		s.preferredExecutionNodeIDs,
+		s.fixedExecutionNodes.NodeIDs(),
+	)
+
+	s.executionResultProviderReal = execution_result.NewExecutionResultInfoProvider(
+		s.log,
+		s.state,
+		s.receipts,
+		executionNodeSelector,
+		s.criteria,
+	)
+}
+
+// TestSubscribeExecutionData verifies that the subscription to execution data works correctly
+// starting from the spork root block ID. It ensures that the execution data is received
+// sequentially and matches the expected data.
+func (s *BackendExecutionDataSuite2) TestSubscribeExecutionData() {
+	s.mockDataProviderState()
+
+	s.executionDataTracker.
+		On("GetStartHeight", mock.Anything, mock.AnythingOfType("flow.Identifier"), mock.AnythingOfType("uint64")).
+		Return(
+			func(ctx context.Context, ID flow.Identifier, height uint64) (uint64, error) {
+				return s.executionDataTrackerReal.GetStartHeight(ctx, ID, height)
+			},
+			nil,
+		).
+		Once()
+
+	backend := s.createExecutionDataBackend()
+	currentHeight := s.sporkRootBlock.Height
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	sub := backend.SubscribeExecutionData(
+		ctx,
+		s.sporkRootBlock.ID(),
+		0, // either id or height must be provided
+		s.criteria,
+	)
+
+	// cancel after we have received all expected blocks to avoid waiting for
+	// streamer/provider to hit the "block not ready" or missing height path.
```

Comment:
do we need to avoid waiting for streamer/provider to hit the "block not ready" or missing height path? if not, don't mention it here otherwise it may confuse the next person to work on these tests since it reads like there is a corner case.
```suggestion
	// cancel after we have received all expected blocks.
```